premerge/post-submit-testing.md

```diff
@@ -0,0 +1,197 @@
+# Post Submit Testing
+
+## Introduction
+
+While this infrastructure is focused on premerge testing, it is also important
+to make sure that the specific configuration we are testing is tested post
+commit as well. This document outlines the motivation for the need to test this
+configuration post commit how we plan on implementing this to ensure we get fast
+feedback scalably, and why we are utilizing this design over others.
+
+## Background/Motivation
+
+LLVM has two types of testing upstream: premerge and postcommit. The premerge
+testing is performed using Github Actions every time a pull request (PR) is
+updated before it is merged. Premerge testing is performed using this
+infrastructure (specifically the `./premerge` folder in llvm-zorg). Landing a PR
+consists of squashing the changes into a single commit and adding that commit to
+the `main` branch in the LLVM monorepo. We care specifically about the state of
+the `main` branch because it is what the community considers to be the canonical
+tree. Currently, commits can also be added to the `main` branch by directly
+pushing to the main branch. Commits pushed directly to `main` are not tested
+through the premerge pipeline as they skip the PR merge process. After a new
+commit lands in the `main` branch, postcommit testing is performed. Most
+postcommit testing is performed through the Buildbot infrastructure. The main
+Buildbot instance for LLVM has a web instance hosted at
+[lab.llvm.org](https://lab.llvm.org/buildbot/#/builders). When a new commit
+lands in `main` the Buildbot instance (sometimes referred to as the Buildbot
+master) will trigger many different builds, base on the configurations
+defined in the llvm-zorg repository under the `buildbot/` folder. These
+configurations are run on Buildbot workers that are hosted by the community.
+Some builders build too slowly to keep up with the pace of commits to `main`,
+so test batches of commits. This often results in a large number of
+erroneous notifications due to the list of possible culprits for a breakage
+being more than a single commit.
+
+For premerge testing, we do not want to notify LLVM developers about failures
+already happening in `main` irrelevant to their changes. This requires knowing
+the state of `main` at the time the premerge testing for a PR was started. We
+also want information on the current state of `main` to empower the community
+with information that they need to revert or forward-fix problematic commits.
+Problematic commits can occur without being caught by the premerge system due to
+someone directly pushing a commit to `main`, or if multiple PRs become
+problematic only when combined. This means we need to test the premerge
+configuration postcommit as well so that we can determine the state of `main`
+(in terms of whether the build passed/failed and what tests failed, if any) at
+any given point in time.
```

Comment:
Recommendation: Add a sentence at the end of this paragraph about the premerge advisor (which you reference in the Data Storage section as if it had been mentioned previously). Something like "We can then use this data to implement a premerge advisor, to help prevent sending false positive test failure notifications to users."